File: src/offer/Questions23.java
package offer;

import offer.LinkedList.Node;

//寻找链表的环节点

public class Questions23 {
	
	public static void main(String[] arg0) {
		Node<String> node1 = new Node<>();
		node1.set("a");
		Node<String> node2 = new Node<>();
		node2.set("b");
		node1.setNext(node2);
		Node<String> node3 = new Node<>();
		node3.set("c");
		node2.setNext(node3);
		Node<String> node4 = new Node<>();
		node4.set("d");
		node3.setNext(node4);
		Node<String> node5 = new Node<>();
		node5.set("e");
		node4.setNext(node5);
		Node<String> node6 = new Node<>();
		node6.set("f");
		node5.setNext(node6);
		node6.setNext(node3);
		System.out.println(findEntry(node1));
	}
	
	public static <T> T findEntry(Node<T> first) {
		int count = isLoop(first);
		if (count == 0) {
			throw new IllegalArgumentException("输入的链表中没有环");
		}
		Node<T> one = first;
		Node<T> two = first;
		for (int i = 1; i <= count; i++) {
			one = one.getNext();
		}
		while (!one.equals(two)) {
			one = one.getNext();
			two = two.getNext();
		}
		return two.get();
	}
	
	private static <T> int isLoop(Node<T> first) {
		Node<T> one = first;
		Node<T> two = first;
		if (one == null || one.getNext() == null) {
			return 0;
		}
		while (one.getNext() != null && one.getNext().getNext() != null) {
			one = one.getNext().getNext();
			two = two.getNext();
			if (one.equals(two)) {
				Node<T> node = one;
				one = one.getNext();
				if (one.equals(node)) {
					return 1;
				}
				int i;
				for (i = 1; !one.equals(node); i++) {
					one = one.getNext();
				}
				return i;
			}
		}
		return 0;
	}

}

File: src/nativeMain/kotlin/com.qiaoyuang.algorithm/round1/Questions23.kt
package com.qiaoyuang.algorithm.round1

fun test23() {
    val node6 = SingleDirectionNode(6)
    val node5 = SingleDirectionNode(5, node6)
    val node4 = SingleDirectionNode(4, node5)
    val node3 = SingleDirectionNode(3, node4)
    val node2 = SingleDirectionNode(2, node3)
    val node1 = SingleDirectionNode(1, node2)
    node6.next = node3
    println("The entry node is: ${node1.findCircleEntryNodeInLinkedList().element}")
}

/**
 * Questions23: Find the circle's entry node in a linked list
 */
private fun <T : Comparable<T>> SingleDirectionNode<T>.findCircleEntryNodeInLinkedList(): SingleDirectionNode<T> {
    if (next == null)
        throw IllegalArgumentException("The linked list doesn't have circle")
    var firstPtr = next!!
    var secondPtr = next!!.next
    while (firstPtr !== secondPtr) {
        firstPtr = firstPtr.next!!
        secondPtr = secondPtr?.next?.next
        if (secondPtr == null)
            throw IllegalArgumentException("The linked list doesn't have circle")
    }
    var circleNodeCount = 1
    secondPtr = secondPtr.next!!
    while (secondPtr !== firstPtr) {
        secondPtr = secondPtr!!.next
        circleNodeCount++
    }
    firstPtr = this
    secondPtr = this
    repeat(circleNodeCount) {
        secondPtr = secondPtr!!.next
    }
    while (firstPtr !== secondPtr) {
        firstPtr = firstPtr.next!!
        secondPtr = secondPtr!!.next!!
    }
    return firstPtr
}